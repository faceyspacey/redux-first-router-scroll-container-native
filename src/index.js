// @flow

// $FlowFixMe - react-native is ignored, which produces an error
import { ScrollView } from 'react-native'
import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import type { Connector } from 'react-redux'

type GenericComponent = Class<React.Component<*, *, *>>

type OwnProps = {
  scrollKey: string,
  component: GenericComponent,
  forceRestore?: boolean,
  onScrollEndDrag?: Event => void,
  onMomentumScrollEnd?: Event => void
}

type StateProps = {
  backNext: ?boolean
}

type Props = OwnProps & StateProps

type XY = {
  x: number,
  y: number,
  animated: boolean
}

type Event = {
  nativeEvent: {
    contentOffset: { x: number, y: number }
  }
}

export class ScrollContainer extends React.Component {
  props: Props

  static scrollViews = {}

  static defaultProps = {
    component: ScrollView
  }

  componentDidMount() {
    this.restoreScroll()
  }

  // `componentDidUpdate` msut also be used in addition to `componentDidMount`
  // since if the `ScrollView` is rendered in the exact same place, the same
  // component instance will be re-used and won't re-mount
  componentDidUpdate({ scrollKey, forceRestore }: Props) {
    if (
      scrollKey !== this.props.scrollKey ||
      (this.props.forceRestore && forceRestore !== this.props.forceRestore)
    ) {
      this.restoreScroll()
    }
  }

  scrollView: Object

  restoreScroll() {
    if (this.props.backNext) {
      const xy = this.getLastXY(this.props)

      if (xy) {
        this.scrollTo(xy)
      }
    }
    else {
      // reset if scroll appears not through a `backNext` action
      // so that if user leaves a 2nd time and returns via `backNext` a stale
      // scroll position is not restored
      this.reset()
    }
  }
  scrollTo(xy: XY) {
    const scrollTo =
      this.scrollView.scrollTo ||
      (this.scrollView._listRef &&
        this.scrollView._listRef._scrollRef.scrollTo) ||
      (this.scrollView._scrollRef && this.scrollView._scrollRef.scrollTo)

    scrollTo
      ? setTimeout(() => scrollTo(xy))
      : console.warn(`[redux-first-router-scroll-container-native] the component
            you passed does not have a 'scrollTo' method. 'FlatList', 'SectionList',
            and 'ScrollView' have been tested. Please submit an issue if you think
            this is a mistake.`)
  }

  getLastXY(): XY {
    return ScrollContainer.scrollViews[this.props.scrollKey]
  }
  setLastXY(e: Event) {
    const { x, y } = e.nativeEvent.contentOffset
    const key = this.props.scrollKey

    ScrollContainer.scrollViews[key] = { x, y, animated: false }
  }
  reset() {
    delete ScrollContainer.scrollViews[this.props.scrollKey]
  }

  onScrollEndDrag = (e: Event) => {
    this.setLastXY(e)
    if (this.props.onScrollEndDrag) {
      this.props.onScrollEndDrag(e)
    }
  }

  onMomentumScrollEnd = (e: Event) => {
    this.setLastXY(e)
    if (this.props.onMomentumScrollEnd) {
      this.props.onMomentumScrollEnd(e)
    }
  }

  render() {
    const {
      scrollKey,
      backNext,
      forceRestore,
      component: ScrollComponent,
      ...props
    } = this.props

    return (
      <ScrollComponent
        {...props}
        ref={node => (this.scrollView = node)}
        key={this.props.scrollKey}
        onScrollEndDrag={this.onScrollEndDrag}
        onMomentumScrollEnd={this.onMomentumScrollEnd}
      />
    )
  }
}

const mapStateToProps = ({ location }) => ({
  backNext: location.backNext
})

const connector: Connector<OwnProps, Props> = connect(mapStateToProps)

const Container = connector(ScrollContainer)

Container.propTypes = {
  scrollKey: PropTypes.string.isRequired,
  component: PropTypes.object, // not required unlike Flow which requires it if you have its defaultProp
  forceRestore: PropTypes.bool,
  onScrollEndDrag: PropTypes.func,
  onMomentumScrollEnd: PropTypes.func
}

export default Container
