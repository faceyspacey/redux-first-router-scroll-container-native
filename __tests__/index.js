// @noflow
import React from 'react'
import { View } from 'react-native'
import renderer from 'react-test-renderer'
import { ScrollContainer } from '../src'

const createScroll = props => (
  <ScrollContainer {...props}>
    <View {...props} />
  </ScrollContainer>
)

test('saves scroll position onMomentumScrollEnd and restores on componentDidMount', () => {
  let ref
  let component
  let instance
  const contentOffset = { x: 10, y: 69, animated: false }

  expect(ScrollContainer.scrollViews).toEqual({})
  const onMomentumScrollEnd = jest.fn()
  component = createScroll({
    ref: node => (ref = node),
    scrollKey: 'foo',
    onMomentumScrollEnd
  })
  instance = renderer.create(component)
  expect(instance.toJSON()).toMatchSnapshot()

  const event = { nativeEvent: { contentOffset } }
  ref.onMomentumScrollEnd(event)
  expect(ScrollContainer.scrollViews.foo).toEqual(contentOffset)
  expect(onMomentumScrollEnd).toHaveBeenCalledWith(event)

  component = createScroll({
    ref: node => (ref = node),
    scrollKey: 'foo',
    backNext: true
  })

  instance = renderer.create(component)

  ref.scrollTo = jest.fn()
  ref.componentDidMount()
  expect(ref.scrollTo).toHaveBeenCalledWith(contentOffset)
})

test('saves scroll position onScrollEndDrag and restores on componentDidUpdate', () => {
  let ref
  let component
  const contentOffset = { x: 10, y: 69, animated: false }

  ScrollContainer.scrollViews = {}
  const onScrollEndDrag = jest.fn()
  component = createScroll({
    ref: node => (ref = node),
    scrollKey: 'foo',
    onScrollEndDrag
  })
  const instance = renderer.create(component)
  expect(instance.toJSON()).toMatchSnapshot()

  const event = { nativeEvent: { contentOffset } }
  ref.onScrollEndDrag(event)
  expect(ScrollContainer.scrollViews.foo).toEqual(contentOffset)
  expect(onScrollEndDrag).toHaveBeenCalledWith(event)

  component = createScroll({
    scrollKey: 'foo',
    backNext: true
  })

  instance.update(component)

  ref.scrollView.scrollTo = jest.fn() // also test actual ScrollView.scrollTo methods
  ref.componentDidUpdate({ scrollKey: 'bar' })
  expect(ref.scrollView.scrollTo).toHaveBeenCalledWith(contentOffset)
})

test('restores scroll on componentDidUpdate when prevProps.forceRestore === false and props.forceRestore is truthy or incremented', () => {
  let ref
  const contentOffset = { x: 10, y: 69, animated: false }
  const component = createScroll({
    ref: node => (ref = node),
    scrollKey: 'foo',
    backNext: true,
    forceRestore: 1
  })

  ScrollContainer.scrollViews = {
    foo: contentOffset
  }

  renderer.create(component)

  ref.scrollTo = jest.fn()
  ref.componentDidUpdate({ scrollKey: 'foo', forceRestore: 0 })
  expect(ref.scrollTo).toHaveBeenCalledWith(contentOffset)
})

test('resets scroll position on mount/update if !props.backNext', () => {
  let ref
  const contentOffset = { x: 10, y: 69, animated: false }
  const component = createScroll({
    ref: node => (ref = node),
    scrollKey: 'foo'
  })

  ScrollContainer.scrollViews = {
    foo: contentOffset
  }

  renderer.create(component)

  // since !props.backNext, scroll position is reset (deleted)
  // so that the scene starts at the top of the page
  ref.reset = jest.fn()
  ref.componentDidMount()
  expect(ref.reset).toHaveBeenCalled()
  expect(ScrollContainer.scrollViews).toEqual({})

  // the scrollKey must change below as `componentDidUpdate` serves the use
  // case where the same component instance is re-used because it appears
  // at the same place in the component tree, but it in fact is representing
  // another scene
  ref.reset = jest.fn()
  ref.scrollTo = jest.fn()
  ref.componentDidUpdate({ scrollKey: 'bar' })
  expect(ref.reset).toHaveBeenCalled()
  expect(ref.scrollTo).not.toHaveBeenCalled()
  expect(ScrollContainer.scrollViews).toEqual({})
})

test('console.warn called when Component without scrollTo method passed as prop', () => {
  let ref
  const contentOffset = { x: 10, y: 69, animated: false }
  const component = createScroll({
    ref: node => (ref = node),
    scrollKey: 'foo',
    backNext: true,
    forceRestore: 1,
    Component: View
  })

  ScrollContainer.scrollViews = {
    foo: contentOffset
  }

  renderer.create(component)

  console.warn = jest.fn()
  ref.componentDidUpdate({ scrollKey: 'foo', forceRestore: 0 })
  expect(console.warn).toHaveBeenCalled()
})
