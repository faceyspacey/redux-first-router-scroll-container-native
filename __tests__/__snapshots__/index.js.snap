// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`saves scroll position onMomentumScrollEnd and restores on componentDidMount 1`] = `
<RCTScrollView
  component={[Function]}
  onMomentumScrollEnd={[Function]}
  onScrollEndDrag={[Function]}
  scrollKey="foo"
>
  <View />
</RCTScrollView>
`;

exports[`saves scroll position onScrollEndDrag and restores on componentDidUpdate 1`] = `
<RCTScrollView
  component={[Function]}
  onMomentumScrollEnd={[Function]}
  onScrollEndDrag={[Function]}
  scrollKey="foo"
>
  <View />
</RCTScrollView>
`;
